### PR TITLE
Test async await

### PR DESCRIPTION
**Pull request by Janaka-Steph**
Rewrite tests with async/await

### CODE REVIEW

**Comment by Janaka-Steph** [comment]
@skmgoldin How is it going ?

**Comment by skmgoldin** [comment]
Can you `git rm` your changes to `package-lock.json`? I don't see a reason why that file should have changed.

File: helpers/index.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+module.exports = {
```
Comment (maurelian) [inline helpers/index.js]: Please move this into the `test` folder, and rename it to utils. 
This'll keep the root clean, and I think utils is a bit more common in truffle.

File: helpers/index.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+  expectThrow: async promise => {
+    try {
+      await promise
+    } catch (error) {
+      // console.warn(error.message)
```
Comment (maurelian) [inline helpers/index.js]: remove this unused console

File: helpers/index.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      // TODO: Check jump destination to destinguish between a throw
```
Comment (maurelian) [inline helpers/index.js]: Please remove this TODO. Recent version of solidity use an `invalid opcode` for throw. https://solidity.readthedocs.io/en/develop/control-structures.html?highlight=exceptions#error-handling-assert-require-revert-and-exceptions

File: helpers/index.js [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+      //       and an actual invalid jump.
+      const invalidJump = error.message.search('invalid JUMP') >= 0
+
+      // TODO: Found out in which circumstances 'invalid opcode' is throwed
```
Comment (maurelian) [inline helpers/index.js]: Minor note, but it should be "Find out .... ".
Also, please remove this TODO per the previous comment.

File: helpers/index.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+      const invalidOpcode = error.message.search('invalid opcode') >= 0
+
+      // TODO: When we contract A calls contract B, and B throws, instead
```
Comment (maurelian) [inline helpers/index.js]: Good observation. This article answers your question: http://truffleframework.com/tutorials/testing-for-throws-in-solidity-tests

Please remove this TODO, if you like, you could add an issue to differentiate between an OOG and failed message call.

File: test/humanStandardToken.js [hunk 6]
```diff
@@ -1,345 +1,186 @@
...
-      assert(true)
-    }).catch((err) => { throw new Error(err) })
+  it('creation: should create an initial balance of 10000 for the creator', async () => {
+    const balance = await HST.balanceOf.call(accounts[0])
+    assert.strictEqual(balance.toNumber(), 10000)
   })
 
-  it('transfers: should transfer 10000 to accounts[1] with accounts[0] having 10000', function () {
-    var ctr
-    return HumanStandardToken.new(10000, 'Simon Bucks', 1, 'SBX', {from: accounts[0]}).then(function (result) {
```
Comment (maurelian) [inline test/humanStandardToken.js]: I like that you've removed all the new contract creations, that was excessive.

**Comment by Janaka-Steph** [comment]
Changes pushed!

**Comment by maurelian** [comment]
Thanks for fast turnaround on that. I'm looking at the rest now. 

I noticed package-lock.json still has changes checked in (https://github.com/ConsenSys/Tokens/pull/71#issuecomment-326422825). Can you summarize what changed?

File: test/humanStandardToken.js [hunk 7]
```diff
@@ -1,345 +1,186 @@
...
-      return ctr.approveAndCall.call(sampleCtr.address, 100, '0x42', {from: accounts[0]})
-    }).catch(function (result) {
-            // It will catch OOG.
+  // this is not *good* enough as the contract could still throw an error otherwise.
+  // ideally one should check balances before and after, but estimateGas currently always throws an error.
+  // it's not giving estimate on gas used in the event of an error.
+  it('transfers: ether transfer should be reversed.', async () => {
+    try {
+      web3.eth.sendTransaction({from: accounts[0], to: HST.address, value: web3.toWei('10', 'Ether')})
+    } catch (e) {
```
Comment (maurelian) [inline test/humanStandardToken.js]: I realize this was not in the in the pre-existing tests, but please check that the balance of the contract is unchanged, then I think we can remove the above comment.

File: test/humanStandardToken.js [hunk 8]
```diff
@@ -1,345 +1,186 @@
...
       assert(true)
```
Comment (maurelian) [inline test/humanStandardToken.js]: With the new utils.js, this could check for a throw via `invalid opcode`.

File: test/humanStandardToken.js [hunk 9]
```diff
@@ -1,345 +1,186 @@
...
+
+  it('approvals: msg.sender should approve 100 to SampleRecipient and then NOTIFY SampleRecipient and throw.', async () => {
+    let SRS = await SampleRecipientThrow.new({from: accounts[0]})
+    return expectThrow(HST.approveAndCall.call(SRS.address, 100, '0x42', {from: accounts[0]}))
+  })
+
+  // bit overkill. But is for testing a bug
+  it('approvals: msg.sender approves accounts[1] of 100 & withdraws 20 once.', async () => {
+    const balance0 = await HST.balanceOf.call(accounts[0])
+    assert.strictEqual(balance0.toNumber(), 10000)
```
Comment (maurelian) [inline test/humanStandardToken.js]: For readability: please add a blank line below this assert, and do the same for other asserts in this function, or in similar tests.

My thinking is that there are several 'mini tests' within this test, so in such cases, some whitespace will help to  divide them. If there are two asserts in a row, they are probably part of the same 'mini test', so don't divide them up.

make sense?

File: test/humanStandardToken.js [hunk 10]
```diff
@@ -1,345 +1,186 @@
...
+    await HST.approve(accounts[1], 100, {from: accounts[0]}) // 100
+    const balance2 = await HST.balanceOf.call(accounts[2])
+    assert.strictEqual(balance2.toNumber(), 0, 'balance2 not correct')
+    // transferFrom(address _from, address _to, uint256 _value)
+    HST.transferFrom.call(accounts[0], accounts[2], 20, {from: accounts[1]})
+    // allowance(address _owner, address _spender)
```
Comment (maurelian) [inline test/humanStandardToken.js]: What is the reason for this comment and the one above? I think they can/should be removed.

File: test/utils.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
+module.exports = {
```
Comment (maurelian) [inline test/utils.js]: This is a very nice addition.

File: test/humanStandardToken.js (hunk 7 above)
Comment (maurelian) [reply test/humanStandardToken.js]: Also, in this catch block, using your new utils.js, this could check for a throw via invalid opcode.

Comment (maurelian) [reply test/humanStandardToken.js]: Yep, while working on another test suite, I found out the error you're getting is probably more like:

{ message: 'Web3ProviderEngine does not support synchronous requests.',
  stack: 'Error: Web3ProviderEngine does not support synchro…/truffle/node_modules/mocha/lib/runner.js:443:10)',
  name: 'ProviderError' }
> e.message.search('invalid
```{ message: 'Web3ProviderEngine does not support synchronous requests.',
  stack: 'Error: Web3ProviderEngine does not support synchro…/truffle/node_modules/mocha/lib/runner.js:443:10)',
  name: 'ProviderError' }```

Comment (Janaka-Steph) [reply test/humanStandardToken.js]: I think there is a misunderstanding here, I didn't wrote this test, it was there before : https://github.com/ConsenSys/Tokens/blob/staging/test/humanStandardToken.js#L55

File: test/humanStandardToken.js (hunk 10 above)
Comment (Janaka-Steph) [reply test/humanStandardToken.js]: As the tests are also used for documentation, I like to add function signature sometimes to avoid having to search in the code. I can remove it if you prefer.

File: test/utils.js (hunk 11 above)
Comment (Janaka-Steph) [reply test/utils.js]: thank you :-)

File: test/humanStandardToken.js (hunk 9 above)
Comment (Janaka-Steph) [reply test/humanStandardToken.js]: yes!

File: test/humanStandardToken.js (hunk 7 above)
Comment (Janaka-Steph) [reply test/humanStandardToken.js]: I had to use `expectThrow` into the sendTransaction callback as this call doesn't return a promise

**Comment by maurelian** [comment]
LGTM, thanks @Janaka-Steph!

**Comment by Janaka-Steph** [comment]
@maurelian I noticed that you reverted my commit #b143e19, even when my work have been merged. Can you explain me why?

**Comment by maurelian** [comment]
Hey @Janaka-Steph, your commit is still in master. I just accidentally clicked the revert button, but that was after the PR was merged. 

You can see in the master branch commit history that  b143e19 is in, but the revert commit aad1d6a is not.

Sorry for the confusion.

**Comment by Janaka-Steph** [comment]
@maurelian Yes I can see the commit if I search the details of ed39594 `Merge latest changes from staging to master (#78)`. But by grouping multiple commits and PR in ed39594 and committing yourself, the result is that I am not considered as a contributor for Github. When opening a test file I should appear as a contributor.

**Comment by maurelian** [comment]
Ahh, yes, I see. That wasn't my intention, and I do want you to be credited. I do want to keep the history clean, so ideally #78 would have had two commits: one commit would be your changes, the second commit would be the small changes I made squashed together. 

Options: 

1. I can try to make that happen by the end of the week. 
2. If you have a chance you're welcome to submit a PR that reverts my squashed commit, and has the two commits described above.

I must reiterate that your help is much appreciated!

**Comment by Janaka-Steph** [comment]
Ok thank you for your quick answer. I vote option 1. Sorry for the little extra work. ^^
Pleasure was mine. :-)

**Comment by Janaka-Steph** [comment]
_friendly reminder_
